### PR TITLE
Background step cancellation thread

### PR DESCRIPTION
**Pull request by johannkm**
## Checklist

- [ ] My change requires a change to the documentation and I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.

### CODE REVIEW

**Comment by johannkm** [comment]
Current dependencies on/for this PR:
* master
  * **PR #5914**  (this PR)

This comment was auto-generated by [Graphite](https://graphite.dev?utm_source=gh) and will continue to be automatically updated while this PR remains open.

File: python_modules/dagster/dagster/cli/api.py [hunk 1]
```diff
@@ -244,51 +244,68 @@ def execute_step_command(input_json):
...
+            finally:
+                if instance.should_start_background_step_thread:
+                    cancellation_thread_shutdown_event.set()
+                    if cancellation_thread.is_alive():
+                        cancellation_thread.join(timeout=15)
+                        if cancellation_thread.is_alive():
+                            instance.report_engine_event(
+                                "Cancellation thread did not shutdown gracefully",
+                                pipeline_run,
+                            )
```
Comment (alangenfeld) [inline python_modules/dagster/dagster/cli/api.py:308]: not enough information to disambiguate

File: python_modules/dagster/dagster/grpc/types.py [hunk 2]
```diff
@@ -124,7 +124,7 @@ class ExecuteStepArgs(
     namedtuple(
         "_ExecuteStepArgs",
         "pipeline_origin pipeline_run_id step_keys_to_execute instance_ref "
-        "retry_mode known_state should_verify_step",
+        "retry_mode known_state should_verify_step should_start_background_cancellation_thread",
```
Comment (alangenfeld) [inline python_modules/dagster/dagster/grpc/types.py:127]: `should_start_background_cancellation_thread` is very current-impl specific. I think from the API perspective the difference is between the run worker being responsible or this step worker being responsible for itself 
riffing:
`self_managed_cancellation: bool`
`cancellation_source: external (default) | self`

File: python_modules/dagster/dagster/cli/api.py [hunk 3]
```diff
@@ -246,51 +249,68 @@ def execute_step_command(input_json):
...
-                click.echo(line)
+                for event in execute_plan_iterator(
+                    execution_plan,
+                    recon_pipeline,
+                    pipeline_run,
+                    instance,
+                    run_config=pipeline_run.run_config,
+                    retry_mode=args.retry_mode,
+                ):
+                    buff.append(serialize_dagster_namedtuple(event))
```
Comment (alangenfeld) [inline python_modules/dagster/dagster/cli/api.py:300]: nit: refactor maybe - nesting getting deep

File: python_modules/dagster/dagster/cli/api.py (hunk 1 above)
Comment (alangenfeld) [reply python_modules/dagster/dagster/cli/api.py:313]: this is attributed to the run not the specific step this is happening in

**Review by alangenfeld — CHANGES_REQUESTED** [review]
See inline comments

**Review by alangenfeld — APPROVED** [review]
🙏

**Comment by alangenfeld** [comment]
this should be put under test right?

at least update the PR summary / test plan with how you tested this

**Pull request closed without merging**